File: src/main/java/jp/co/cybird/barcodekanojoForGAM/adapter/DashboardAdapter.java
package jp.co.cybird.barcodekanojoForGAM.adapter;

import android.content.Context;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import java.util.Iterator;
import java.util.Observable;
import java.util.Observer;
import jp.co.cybird.barcodekanojoForGAM.R;
import jp.co.cybird.barcodekanojoForGAM.adapter.base.BaseActivityModelAdapter;
import jp.co.cybird.barcodekanojoForGAM.adapter.base.ObservableAdapter;
import jp.co.cybird.barcodekanojoForGAM.core.model.ActivityModel;
import jp.co.cybird.barcodekanojoForGAM.core.model.Kanojo;
import jp.co.cybird.barcodekanojoForGAM.core.model.ModelList;
import jp.co.cybird.barcodekanojoForGAM.core.util.ImageCache;
import jp.co.cybird.barcodekanojoForGAM.core.util.RemoteResourceManager;

public class DashboardAdapter extends BaseActivityModelAdapter implements ObservableAdapter {
    private static final boolean DEBUG = false;
    private static final String TAG = "DashboardAdapter";
    private boolean isFirst = false;
    private Handler mHandler;
    private LayoutInflater mInflater;
    private OnKanojoClickListener mListener;
    private int mLoadedPhotoIndex = 0;
    private final Runnable mNotifyThread = new Runnable() {
        public void run() {
            DashboardAdapter.this.superNotifyDataSetChanged();
        }
    };
    private RemoteResourceManagerObserver mResourcesObserver;
    private RemoteResourceManager mRrm;
    private Runnable mRunnableLoadPhotos = new Runnable() {
        public void run() {
            if (DashboardAdapter.this.mLoadedPhotoIndex < DashboardAdapter.this.getCount()) {
                DashboardAdapter dashboardAdapter = DashboardAdapter.this;
                DashboardAdapter dashboardAdapter2 = DashboardAdapter.this;
                int access$0 = dashboardAdapter2.mLoadedPhotoIndex;
                dashboardAdapter2.mLoadedPhotoIndex = access$0 + 1;
                ActivityModel a = (ActivityModel) dashboardAdapter.getItem(access$0);
                ImageCache.requestImage(a.getLeftImgUrl(), DashboardAdapter.this.mRrm);
                ImageCache.requestImage(a.getRightImgUrl(), DashboardAdapter.this.mRrm);
                DashboardAdapter.this.mHandler.postDelayed(DashboardAdapter.this.mRunnableLoadPhotos, 200);
            }
        }
    };

    public interface OnKanojoClickListener {
        void onKanojoClick(Kanojo kanojo);
    }

    public DashboardAdapter(Context context, RemoteResourceManager rrm, Observer observer) {
        super(context);
        this.mInflater = LayoutInflater.from(context);
        this.mRrm = rrm;
        this.mHandler = new Handler();
        this.mResourcesObserver = new RemoteResourceManagerObserver(this, (RemoteResourceManagerObserver) null);
        this.mRrm.addObserver(this.mResourcesObserver);
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder holder;
        View view = convertView;
        if (view == null) {
            view = this.mInflater.inflate(R.layout.row_activities, null);
            holder = new ViewHolder(this, null);
            holder.imgLeft = view.findViewById(R.id.row_activities_left_img);
            holder.imgRight = view.findViewById(R.id.row_activities_right_img);
            holder.imgRightCover = view.findViewById(R.id.row_activities_right_img_cover);
            holder.txtActivity = view.findViewById(R.id.row_activities_txt);
            view.setTag(holder);
        } else {
            holder = (ViewHolder) view.getTag();
        }
        ActivityModel act = (ActivityModel) getItem(position);
        if (act != null) {
            ImageCache.setImage(holder.imgLeft, act.getLeftImgUrl(), this.mRrm, R.drawable.common_noimage_product);
            ImageCache.setImage(holder.imgRight, act.getRightImgUrl(), this.mRrm, R.drawable.common_noimage_product);
            if (holder.txtActivity != null) {
                holder.txtActivity.setText(act.getActivity());
            }
            final Kanojo kanojo = act.getKanojo();
            holder.imgLeft.setOnClickListener(null);
            holder.imgRight.setOnClickListener(null);
            holder.imgRight.setVisibility(View.VISIBLE);
            holder.imgRightCover.setVisibility(View.VISIBLE);
            switch (act.getActivity_type()) {
                case 2:
                case 9:
                    holder.imgLeft.setOnClickListener(new View.OnClickListener() {
                        public void onClick(View v) {
                            if (DashboardAdapter.this.mListener != null) {
                                DashboardAdapter.this.mListener.onKanojoClick(kanojo);
                            }
                        }
                    });
                    break;
                case 5:
                case 7:
                case 8:
                case 10:
                    holder.imgRight.setOnClickListener(new View.OnClickListener() {
                        public void onClick(View v) {
                            if (DashboardAdapter.this.mListener != null) {
                                DashboardAdapter.this.mListener.onKanojoClick(kanojo);
                            }
                        }
                    });
                    break;
                case 11:
                    holder.imgRight.setVisibility(View.GONE);
                    holder.imgRightCover.setVisibility(View.GONE);
                    break;
            }
        }
        return view;
    }

    public void setOnKanojoClickListener(OnKanojoClickListener l) {
        this.mListener = l;
    }

    public void setModelList(ModelList<ActivityModel> l) {
        super.setModelList(l);
        this.mHandler.removeCallbacks(this.mRunnableLoadPhotos);
        this.mHandler.postDelayed(this.mRunnableLoadPhotos, 100);
    }

    public void addModelList(ModelList<ActivityModel> l) {
        super.addModelList(l);
        Iterator it = l.iterator();
        while (it.hasNext()) {
            ActivityModel a = (ActivityModel) it.next();
            ImageCache.requestImage(a.getLeftImgUrl(), this.mRrm);
            ImageCache.requestImage(a.getRightImgUrl(), this.mRrm);
        }
    }

    public void removeObserver() {
        this.mHandler.removeCallbacks(this.mNotifyThread);
        this.mHandler.removeCallbacks(this.mRunnableLoadPhotos);
        this.mRrm.deleteObserver(this.mResourcesObserver);
    }

    private class RemoteResourceManagerObserver implements Observer {
        private RemoteResourceManagerObserver() {
        }

        /* synthetic */ RemoteResourceManagerObserver(DashboardAdapter dashboardAdapter, RemoteResourceManagerObserver remoteResourceManagerObserver) {
            this();
        }

        public void update(Observable observable, Object data) {
            if (data != null) {
                DashboardAdapter.this.mHandler.removeCallbacks(DashboardAdapter.this.mNotifyThread);
                DashboardAdapter.this.mHandler.postDelayed(DashboardAdapter.this.mNotifyThread, 400);
            }
        }
    }

    public void notifyDataSetInvalidated() {
        super.notifyDataSetInvalidated();
    }

    public void notifyDataSetChanged() {
        super.notifyDataSetChanged();
        this.mHandler.removeCallbacks(this.mNotifyThread);
        this.mHandler.postDelayed(this.mNotifyThread, 400);
    }

    public void superNotifyDataSetChanged() {
        if (this.isFirst) {
            super.notifyDataSetInvalidated();
        } else {
            super.notifyDataSetChanged();
        }
    }

    private class ViewHolder {
        ImageView imgLeft;
        ImageView imgRight;
        ImageView imgRightCover;
        TextView txtActivity;

        private ViewHolder() {
        }

        /* synthetic */ ViewHolder(DashboardAdapter dashboardAdapter, ViewHolder viewHolder) {
            this();
        }
    }
}

File: src/main/java/com/goujer/barcodekanojo/adapter/DashboardAdapter.kt
package com.goujer.barcodekanojo.adapter

import android.content.Context
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.goujer.barcodekanojo.core.util.DynamicImageCache
import jp.co.cybird.barcodekanojoForGAM.R
import jp.co.cybird.barcodekanojoForGAM.adapter.base.BaseActivityModelAdapter
import jp.co.cybird.barcodekanojoForGAM.adapter.base.ObservableAdapter
import jp.co.cybird.barcodekanojoForGAM.core.model.ActivityModel
import jp.co.cybird.barcodekanojoForGAM.core.model.Kanojo
import jp.co.cybird.barcodekanojoForGAM.core.model.ModelList
import kotlinx.coroutines.*
import java.lang.Runnable
import java.util.*

class DashboardAdapter(context: Context?, private val mDic: DynamicImageCache, observer: Observer?) : BaseActivityModelAdapter(context), ObservableAdapter {
	private val isFirst = false
	private val mHandler = Handler()
	private val mInflater = LayoutInflater.from(context)
	private var mListener: OnKanojoClickListener? = null
	private var mLoadedPhotoIndex = 0
	private val mNotifyThread = Runnable { superNotifyDataSetChanged() }
	private val mResourcesObserver = RemoteResourceManagerObserver(this, null as RemoteResourceManagerObserver?)
	private val mRunnableLoadPhotos: Runnable = object : Runnable {
		override fun run() {
			if (mLoadedPhotoIndex < this@DashboardAdapter.count) {
				val dashboardAdapter = this@DashboardAdapter
				val dashboardAdapter2 = this@DashboardAdapter
				val access0 = dashboardAdapter2.mLoadedPhotoIndex
				dashboardAdapter2.mLoadedPhotoIndex = access0 + 1
				val a = dashboardAdapter.getItem(access0) as ActivityModel
				mDic.requestBitmap(a.leftImgUrl)
				mDic.requestBitmap(a.rightImgUrl)
				mHandler.postDelayed(this, 200)
			}
		}
	}
	private val mScope = MainScope()

	fun destroy() {
		mScope.cancel()
	}

	interface OnKanojoClickListener {
		fun onKanojoClick(kanojo: Kanojo?)
	}

	override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {
		val holder: ViewHolder
		var view = convertView
		if (view == null) {
			view = mInflater.inflate(R.layout.row_activities, null)
			holder = ViewHolder()
			holder.imgLeft = view.findViewById(R.id.row_activities_left_img)
			holder.imgRight = view.findViewById(R.id.row_activities_right_img)
			holder.imgRightCover = view.findViewById(R.id.row_activities_right_img_cover)
			holder.txtActivity = view.findViewById(R.id.row_activities_txt)
			view.tag = holder
		} else {
			holder = view.tag as ViewHolder
			holder.reset()
		}
		val act = getItem(position) as ActivityModel
		if (act != null) {
			holder.leftImageJob = mScope.launch { mDic.loadBitmap(holder.imgLeft!!, act.leftImgUrl, R.drawable.common_noimage_product, null) }
			holder.rightImageJob = mScope.launch { mDic.loadBitmap(holder.imgRight!!, act.rightImgUrl, R.drawable.common_noimage_product, null) }
			if (holder.txtActivity != null) {
				holder.txtActivity!!.text = act.activity
			}
			val kanojo = act.kanojo
			holder.imgLeft!!.setOnClickListener(null)
			holder.imgRight!!.setOnClickListener(null)
			holder.imgRight!!.visibility = View.VISIBLE
			holder.imgRightCover!!.visibility = View.VISIBLE
			when (act.activity_type) {
				2, 9 -> holder.imgLeft!!.setOnClickListener {
					if (mListener != null) {
						mListener!!.onKanojoClick(kanojo)
					}
				}
				5, 7, 8, 10 -> holder.imgRight!!.setOnClickListener {
					if (mListener != null) {
						mListener!!.onKanojoClick(kanojo)
					}
				}
				11 -> {
					holder.imgRight!!.visibility = View.GONE
					holder.imgRightCover!!.visibility = View.GONE
				}
			}
		}
		return view
	}

	fun setOnKanojoClickListener(l: OnKanojoClickListener?) {
		mListener = l
	}

	override fun setModelList(l: ModelList<ActivityModel>) {
		super.setModelList(l)
		mHandler.removeCallbacks(mRunnableLoadPhotos)
		mHandler.postDelayed(mRunnableLoadPhotos, 100)
	}

	override fun addModelList(l: ModelList<ActivityModel>) {
		super.addModelList(l)
		for (a in l) {
			mDic.requestBitmap(a.leftImgUrl)
			mDic.requestBitmap(a.rightImgUrl)
		}
	}

	override fun removeObserver() {
		mHandler.removeCallbacks(mNotifyThread)
		mHandler.removeCallbacks(mRunnableLoadPhotos)
		mDic.deleteObserver(mResourcesObserver)
	}

	private inner class RemoteResourceManagerObserver private constructor() : Observer {
		/* synthetic */
		internal constructor(dashboardAdapter: DashboardAdapter?, remoteResourceManagerObserver: RemoteResourceManagerObserver?) : this() {}

		override fun update(observable: Observable, data: Any) {
			if (data != null) {
				mHandler.removeCallbacks(mNotifyThread)
				mHandler.postDelayed(mNotifyThread, 400)
			}
		}
	}

	override fun notifyDataSetChanged() {
		super.notifyDataSetChanged()
		mHandler.removeCallbacks(mNotifyThread)
		mHandler.postDelayed(mNotifyThread, 400)
	}

	fun superNotifyDataSetChanged() {
		if (isFirst) {
			super.notifyDataSetInvalidated()
		} else {
			super.notifyDataSetChanged()
		}
	}

	private class ViewHolder {
		var imgLeft: ImageView? = null
		var imgRight: ImageView? = null
		var imgRightCover: ImageView? = null
		var txtActivity: TextView? = null
		var rightImageJob: Job? = null
		var leftImageJob: Job? = null
		fun reset() {
			if (leftImageJob != null && !leftImageJob!!.isCompleted) {
				leftImageJob!!.cancel(null)
			}
			if (rightImageJob != null && !rightImageJob!!.isCompleted) {
				rightImageJob!!.cancel(null)
			}
		}
	}

	companion object {
		private const val TAG = "DashboardAdapter"
	}

	init {
		mDic.addObserver(mResourcesObserver)
	}
}